pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -311,20 +340,46 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 
 	cidrStrings := make([]string, 0)
 
-	if len(instance.NetworkInterfaces) == 0 || (len(instance.NetworkInterfaces) == 1 && len(instance.NetworkInterfaces[0].AliasIpRanges) == 0) {
+	// No Pod CIDRs can be allocated to a node, if there are no network interfaces;
+	// or if there is one network interface, but it has neither an IP alias nor an IPv6 address.
+	if len(instance.NetworkInterfaces) == 0 ||
+		(len(instance.NetworkInterfaces) == 1 &&
+			len(instance.NetworkInterfaces[0].AliasIpRanges) == 0 &&
+			ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]) == nil) {
+
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: Node %v has no ranges from which CIDRs can be allocated", node.Name)
 	}
 
-	// sets the v1.NodeNetworkUnavailable condition to False
+	// Sets the v1.NodeNetworkUnavailable condition to False.
 	ca.setNetworkCondition(node)
 
-	// nodes in clusters WITHOUT multi-networking are expected to have only 1 network-interface with 1 alias IP range.
-	if len(instance.NetworkInterfaces) == 1 && len(instance.NetworkInterfaces[0].AliasIpRanges) == 1 {
-		cidrStrings = append(cidrStrings, instance.NetworkInterfaces[0].AliasIpRanges[0].IpCidrRange)
-		ipv6Addr := ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0])
-		if ipv6Addr != nil {
-			cidrStrings = append(cidrStrings, ipv6Addr.String())
+	// Nodes in clusters WITHOUT multi-networking are expected to have only 1 network-interface
+	// with 1 alias IPv4 range and/or 1 IPv6 address. Multi-network clusters may have 1 interface
+	// with multiple aliases.
+	if len(instance.NetworkInterfaces) == 1 &&
+		(len(instance.NetworkInterfaces[0].AliasIpRanges) == 1 ||
+			ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]) != nil) {
+
+		ipv4CIDR := ""
+		if len(instance.NetworkInterfaces[0].AliasIpRanges) > 0 {
+			ipv4CIDR = instance.NetworkInterfaces[0].AliasIpRanges[0].IpCidrRange
+		}
+
+		ipv6CIDR := ""
+		if addr := ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]); addr != nil {
+			ipv6CIDR = addr.String()
+		}
+
+		switch ca.stackType {
+		case stackIPv4:
+			cidrStrings = []string{ipv4CIDR}
+		case stackIPv4IPv6:
+			cidrStrings = []string{ipv4CIDR, ipv6CIDR}
+		case stackIPv6IPv4:
+			cidrStrings = []string{ipv6CIDR, ipv4CIDR}
+		case stackIPv6:
+			cidrStrings = []string{ipv6CIDR}
```

Comment:
I still think we should validate here
```suggestion
		switch ca.stackType {
		case stackIPv4 && ipv4CIDR != "" :
			cidrStrings = []string{ipv4CIDR}
		case stackIPv4IPv6 && ipv4CIDR != "" && ipv6CIDR != "":
			cidrStrings = []string{ipv4CIDR, ipv6CIDR}
		case stackIPv6IPv4 && ipv4CIDR != "" && ipv6CIDR != "":
			cidrStrings = []string{ipv6CIDR, ipv4CIDR}
		case stackIPv6 && ipv6CIDR != "":
			cidrStrings = []string{ipv6CIDR}
	         default:
	              return fmt.Errorf("failed to allocate cidr: Node %v has no ranges from which CIDRs can be allocated for the cluster stack family %s", node.Name, ca.stackType)
```

Reply:
Thanks - done.